pysam/libcbcf.pyi

```diff
@@ -0,0 +1,457 @@
+import sys
+from typing import (
+    Optional,
+    Union,
+    Any,
+    Sequence,
+    Tuple,
+    Iterator,
+    List,
+    Iterable,
+    Dict,
+    overload,
+    TypeVar,
+    Mapping,
+)
+
+if sys.version_info < (3, 8):
+    from typing_extensions import Literal
+else:
+    from typing import Literal
+
+from pysam.libchtslib import HTSFile
+
+_D = TypeVar("_D")
+
+class VariantHeaderRecord:
```

Comment:
Should also be Mapping[str, Optional[str]]?

Reply:
It isn't immediately clear why value should be Optional, but it has been a long time since I've dug into this code.

Reply:
Yes, this should definitely be simplified. I tried to factor out a couple of base classes, but then noticed subtle differences where the types didn’t quite match. For example, your `keys()` method returns a Python-2-style list, but `Mapping.keys()` returns an iterator. Perhaps it works anyway, but I decided to spell everything out explicitly for now. That was necessary as a first step anyway. 

The value can be None, see `VariantHeaderRecord.itervalues()`:
https://github.com/pysam-developers/pysam/blob/e340c8574694ab16fbe290c6fb714c1f6c2f5664/pysam/libcbcf.pyx#L1335-L1343

Reply:
This was all written for Python 2, so it doesn't surprise me that it was never updated to the new ABCs.  Sounds like work that should be done in a follow-on PR.  Same for figuring out why keys can be None in `VariantHeaderRecord`s.

Reply:
I have pushed an updated version that uses a custom `_Mapping` class for all the common methods.

When I try to subclass one of the ABCs like this:
```
class VariantHeaderRecord(Sized):
```
I get the error message that I haven’t implemented all abstract methods (`__len__` in this case), so it doesn’t actually help me reduce the number of lines.